js/fetch.ts

```diff
@@ -20,10 +20,265 @@ import {
+  for (let i = 0; i < len; ++i) {
+    const byte = value.charCodeAt(i);
+    if (byte === 0x00 || byte === 0x0a || byte === 0x0d) {
+      return false;
+    }
+  }
+  return true;
+}
+
+function isForbiddenHeaderName(name: string): boolean {
+  return (
+    [
+      "accept-charset",
+      "accept-encoding",
+      "access-control-request-headers",
+      "access-control-request-method",
+      "connection",
+      "content-length",
+      "cookie",
+      "cookie2",
+      "date",
+      "dnt",
+      "expect",
+      "host",
+      "keep-alive",
+      "origin",
+      "referer",
+      "te",
+      "trailer",
+      "transfer-encoding",
+      "upgrade",
+      "via"
+    ].indexOf(name.toLowerCase()) > -1
+  );
+}
+// https://fetch.spec.whatwg.org/#cors-safelisted-request-header
+function isCorsSafelistedRequestHeader(header: Header): boolean {
+  const value = header.value;
+  switch (header.name.toLowerCase()) {
+    case "accept":
+      for (let i = 0; i < value.length; ++i) {
+        // Value contains a byte less than 0x20 that is not 0x09 HT.
+        const byte = value.charCodeAt(i);
+        if (byte < 0x20 && byte !== 0x09) {
+          return false;
+        }
+        // Value contains 0x22 ("), 0x28 (left parenthesis),
+        // 0x29 (right parenthesis), 0x3A (:), 0x3C (<), 0x3E (>),
+        // 0x3F (?), 0x40 (@), 0x5B ([), 0x5C (\), 0x5D (]), 0x7B ({),
+        // 0x7D (}), or 0x7F DEL
+        if (
+          byte === 0x22 ||
+          byte === 0x28 ||
+          byte === 0x29 ||
+          byte === 0x3a ||
+          byte === 0x3c ||
+          byte === 0x3e ||
+          byte === 0x3f ||
+          byte === 0x40 ||
+          byte === 0x5b ||
+          byte === 0x5c ||
+          byte === 0x5d ||
+          byte === 0x7b ||
+          byte === 0x7d ||
+          byte === 0x7f
+        ) {
+          return false;
+        }
+      }
+      break;
+    case "accept-language":
+    case "content-language":
+      // If value contains a byte that is not in the range 0x30 (0) to 0x39 (9)
+      // inclusive, is not in the range 0x41 (A) to 0x5A (Z), inclusive, is not
+      // in the range 0x61 (a) to 0x7A (z), inclusive, and is not 0x20 (SP),
+      // 0x2A (*), 0x2C (,), 0x2D (-), 0x2E (.), 0x3B (;), or 0x3D (=), then
+      // return false.
+      for (let i = 0; i < value.length; ++i) {
+        const byte = value.charCodeAt(i);
+        if (
+          !(0x30 <= byte && byte <= 0x39) &&
+          !(0x41 <= byte && byte <= 0x5a) &&
+          !(0x61 <= byte && byte <= 0x7a) &&
+          byte !== 0x20 &&
+          byte !== 0x2a &&
+          byte !== 0x2c &&
+          byte !== 0x2d &&
+          byte !== 0x2e &&
+          byte !== 0x3b &&
+          byte !== 0x3d
+        ) {
+          return false;
+        }
+      }
+      break;
+    // TODO(qti3e) DRP, DownloadLink, ....
+    default:
+      return false;
+  }
+  if (value.length > 128) {
+    return false;
+  }
+  return true;
+}
```

Comment:
I don’t think it’s necessary at this stage to be fully compliant with web fetch - but I appreciate the effort. If we’re to land this code each of these functions should have unit tests.